### PR TITLE
fix(weave): Improve core arrow stability

### PR DESCRIPTION
**Pull request by shawnlewis**
This fixes a number of issues found by our arrow hypothesis (fuzz) tests. Those tests cover:
- saving and loading weave data as arrow
- concatenation
- joining
- to_compare_safe (used in sort, groupby, join)

I can raise EXAMPLES_PER_TEST to 10,000 locally, and the hypothesis tests will succeed without error. We run with just 100 in CI, so things should be much more stable.

Introduces a new format for saved arrow tables, with backward compatibility:
- we now save ArrowWeaveList in arrow's native feather format, instead of parquet. A nice side effect is that feather is much faster to read, while being similar in size for realistic data.
- there were issues that were hard to fix in our to/from parquet conversion, using feather means we don't need to convert
- the new format also lifts saved structs to the top level of table (when the type is TypedDict), so we can read specific columns out efficiently (reading subset of columns will be in a subsequent PR)
- we preserve backward compat with previously saved data by storing a version number in the artifact metadata

Fixes the case where a callback passed to map columns returns a reducible (by our merge_types rules) union, which fixes a number of edge cases triggered by the hypothesis test.

There are a a few other small bug fixes for various edge cases.

### CODE REVIEW

File: weave/tests/test_hypothesis.py [hunk 1]
```diff
@@ -238,29 +260,27 @@ def test_to_compare_safe(l1):
...
     compare_py(l1_result, l1)
     a1 = ops_arrow.to_arrow(l1)
-
-    # There are bugs here that hypothesis finds! Commenting out for now.
-    # a1_ref = storage.save(a1)
-    # a1_again = storage.get(str(a1_ref))
-    # assert a1.object_type == a1_again.object_type
-    # assert a1._arrow_data == a1_again._arrow_data
-    # a1 = a1_again
-
```
Comment (dannygoldstein) [inline weave/tests/test_hypothesis.py:255]: does this PR not fix these? should we re-enable?

File: weave/tests/test_arrow.py [hunk 2]
```diff
@@ -1654,218 +1655,17 @@ def test_save_nested_custom_objs():
...
-    assert weave.use(awl_node).to_pylist_raw() == data
-
-    # Assert concat-ability
-    raw_concat_awl = arrow.to_arrow(
-        concat_with_data, types.List(concat_with_weave_type)
-    )
-    concat_awl = weave.save(raw_concat_awl)
-    concatted = ops.make_list(a=awl_node, b=concat_awl).concat()
-    assert weave.use(concatted).to_pylist_raw() == data + concat_with_data
-
```
Comment (dannygoldstein) [inline weave/tests/test_arrow.py:1865]: why are we removing this?

File: weave/ops_arrow/arrow.py [hunk 3]
```diff
@@ -182,31 +183,78 @@ def save_instance(self, obj, artifact, name):
...
+        # with artifact.new_file(f"{name}.ArrowWeaveList.parquet", binary=True) as f:
+        #     pq.write_table(table, f)
+
+        # v2 AWL format
+        # - if it's a top-level typeddict we don't save in nested sub-array.
+        #   reading specific columns from a parquet or feather file only gives a perf
+        #   improvement for top-level columns.
+        # - saved as feather file instead of parquet. There are some issues in our
+        #   to/from parquet conversion with unions. They can be triggered by using
+        #   v1 format with the hypothesis tests.
```
Comment (dannygoldstein) [inline weave/ops_arrow/arrow.py:200]: does feather have full support for all arrow formats (e.g., can I directly write a union of dictionary encoded structs)? or do we still need to do massaging for certain exotic arrow arrays?

File: weave/ops_arrow/arrow.py [hunk 4]
```diff
@@ -182,31 +183,78 @@ def save_instance(self, obj, artifact, name):
...
             )
 
-        from . import convert
+        # v1 AWL format
+        # from . import convert
+
+        # parquet_friendly = convert.to_parquet_friendly(obj)
+        # table = pa.table({"arr": parquet_friendly._arrow_data})
+        # with artifact.new_file(f"{name}.ArrowWeaveList.parquet", binary=True) as f:
+        #     pq.write_table(table, f)
```
Comment (dannygoldstein) [inline weave/ops_arrow/arrow.py:192]: do we still support reading existing parquet files after this? or will we have to wipe the cache and transition over fully to feather?

File: weave/ops_arrow/arrow.py (hunk 3 above)
Comment (shawnlewis) [reply weave/ops_arrow/arrow.py:200]: Feather is literally the arrow memory format, but on disk, so it should support all arrow features!

File: weave/ops_arrow/arrow.py (hunk 4 above)
Comment (shawnlewis) [reply weave/ops_arrow/arrow.py:192]: The read side implements both versions. So we don't need to wipe the cache.

File: weave/tests/test_arrow.py (hunk 2 above)
Comment (shawnlewis) [reply weave/tests/test_arrow.py:1865]: This tests generates a bunch of invalid states that we can't really get to anymore, so I removed it.

File: weave/tests/test_hypothesis.py (hunk 1 above)
Comment (shawnlewis) [reply weave/tests/test_hypothesis.py:255]: Because I moved this out into a separate test.

**Review by dannygoldstein — APPROVED** [review]
